tests/testthat/test-expr-string.R

```diff
@@ -0,0 +1,931 @@
+#     pl$lit(c("foo bar", "foo-bar", "foo bar baz"))$str$split(by = " ")$to_r(),
+#     list(c("foo", "bar"), "foo-bar", c("foo", "bar", "baz"))
+#   )
+
+#   expect_identical(
+#     pl$lit(c("foo bar", "foo-bar", "foo bar baz"))$str$split(by = " ", inclusive = TRUE)$to_r(),
+#     list(c("foo ", "bar"), "foo-bar", c("foo ", "bar ", "baz"))
+#   )
+
+#   expect_identical(
+#     pl$lit(c("foo bar", "foo-bar", "foo bar baz"))$str$split(by = "-", inclusive = TRUE)$to_r(),
+#     list("foo bar", c("foo-", "bar"), "foo bar baz")
+#   )
+
+#   expect_grepl_error(
+#     pl$DataFrame(pl$lit("42")$str$split(by = 42L, inclusive = TRUE)),
+#     "invalid series dtype"
+#   )
+
+#   expect_grepl_error(
+#     pl$lit("42")$str$split(by = "blop", inclusive = 42),
+#     "bool"
+#   )
+
+#   # with expression in "by" arg
+#   df <- pl$DataFrame(s = c("foo^bar", "foo_bar", "foo*bar*baz"), "by" = c("_", "_", "*"))
+#   expect_identical(
+#     df$select(pl$col("s")$str$split(by = pl$col("by"))) |> as.list()[[1]],
+#     list("foo^bar", c("foo", "bar"), c("foo", "bar", "baz"))
+#   )
+# })
+
+# test_that("str$split_exact", {
+#   expect_identical(
+#     pl$lit(c("foo bar", "bar foo", "foo bar baz"))$str$split_exact(by = " ", n = 1)$to_r(),
+#     structure(list(field_0 = c("foo", "bar", "foo"), field_1 = c("bar", "foo", "bar")), is_struct = TRUE)
+#   )
+
+#   expect_identical(
+#     pl$lit(c("foo bar", "bar foo", "foo bar baz"))$str$split_exact(by = " ", n = 2)$to_r(),
+#     structure(list(field_0 = c("foo", "bar", "foo"), field_1 = c("bar", "foo", "bar"), field_2 = c(NA, NA, "baz")), is_struct = TRUE)
+#   )
+
+#   expect_identical(
+#     pl$lit(c("foo bar", "foo-bar", "foo bar baz"))$str$split(by = "-", inclusive = TRUE)$to_r(),
+#     list("foo bar", c("foo-", "bar"), "foo bar baz")
+#   )
+
+#   expect_grepl_error(
+#     pl$lit("42")$str$split_exact(by = "a", n = -1, inclusive = TRUE),
+#     "cannot be less than zero"
+#   )
+
+#   expect_grepl_error(
+#     pl$lit("42")$str$split_exact(by = "a", n = 2, inclusive = "joe"),
+#     "bool"
+#   )
+# })
+
+
+# test_that("str$split_exact", {
+#   expect_identical(
+#     pl$lit(c("a_1", NA, "c", "d_4-5"))$str$splitn(by = "_", 1)$to_r(),
+#     structure(list(field_0 = c("a_1", NA, "c", "d_4-5")), is_struct = TRUE)
+#   )
+
+#   expect_identical(
+#     pl$lit(c("a_1", NA, "c", "d_4-5"))$str$splitn(by = "_", 2)$to_r(),
+#     structure(list(field_0 = c("a", NA, "c", "d"), field_1 = c("1", NA, NA, "4-5")), is_struct = TRUE)
+#   )
+
+#   expect_identical(
+#     pl$lit(c("a_1", NA, "c", "d_4-5"))$str$splitn(by = "-", 2)$to_r(),
+#     structure(list(
+#       field_0 = c("a_1", NA, "c", "d_4"),
+#       field_1 = c(NA, NA, NA, "5")
+#     ), is_struct = TRUE)
+#   )
+# })
+
+
+test_that("str$replace", {
+  dat <- pl$DataFrame(x = c("123abc", "abc456"))
+  expect_identical(
+    dat$with_columns(pl$col("x")$str$replace(r"{abc\b}", "ABC")) |>
+      as.list(),
+    list(x = c("123ABC", "abc456"))
+  )
```

Comment:
Unlike current r-polars, polars_data_frame can be compared directly with `expect_equal()`, so do not use `as.list()`.
https://github.com/eitsupi/neo-r-polars/blob/d5994dc6a55d0cdea27a77eeb54fa04be701c1fd/tests/testthat/test-as_polars_df.R#L35-L38

Reply:
Is `expect_equal()` redefined somewhere? I don't understand how it accepts `polars` objects? Also, it converts i64 values to character, so this breaks in this case (but not if I use `as.list()`):
```
dat <- pl$DataFrame(x = c("110", "101", "010"))
expect_equal(
  dat$with_columns(pl$col("x")$str$to_integer(base = 2)),
  pl$DataFrame(x = c(6, 5, 2))
)

#> `actual$x` is a character vector ('6', '5', '2')
#> `expected$x` is a double vector (6, 5, 2)
```

Reply:
Hmmm, that is quite odd.

The comparison is defined in the S3 method of `waldo::compare()`.
https://github.com/eitsupi/neo-r-polars/blob/800cdb866057b7f45292119b2b03ba99eed474c8/R/dataframe-s3-waldo.R

Reply:
> Also, it converts i64 values to character, so this breaks in this case

Sorry, I forgot that was defined here. Could you cast the expected and match the dtype?
https://github.com/eitsupi/neo-r-polars/blob/800cdb866057b7f45292119b2b03ba99eed474c8/R/series-s3-waldo.R#L5

Reply:
Thanks, it works like that (example above still fails because original is int and expected is float but that's expected)